p2p/transport/quic/transport.go

```diff
@@ -38,17 +39,116 @@ var quicDialContext = quic.DialContext // so we can mock it in tests
 
 var HolePunchTimeout = 5 * time.Second
 
+// The higher the value, the more accurate the result, but more resources will be used
+const bufferLength = 3
+
+type bufferEntry struct {
+	failedRetries, retries, total uint
+}
+
+type accept struct {
+	buffer           [bufferLength]bufferEntry
+	mtx              sync.Mutex
+	duration         time.Duration
+	fraction         float64
+	start            sync.Once
+	startTime        time.Time // Time of the first execution
+	lastCycle        int64     // last period id (incremental)
+	retryProbability float64   // probability for the Retry packet in the actual time window
+}
+
+func (a *accept) setOptions(duration time.Duration, fraction float64) *accept {
+	if duration == 0 {
+		duration = 5 * time.Minute
+	}
+
+	if fraction == 0 {
+		fraction = 0.5
+	}
+
+	a.duration = duration
+	a.fraction = fraction
+	return a
+}
+
+func (a *accept) acceptFunction(clientAddr net.Addr, token *quic.Token) bool {
+	a.start.Do(func() { a.startTime = time.Now() })
```

Comment:
This could also be done in the constructor, then you don't need a `sync.Once`.

Reply:
In the first implementation I did something like that, but I don't know the time difference between the object creation and the first invocation, if the difference is negligible I can modify